# Thomas Brewer - Project 1 : Pokemon Stay

This is my project submission, with all questions answered.  Some questions I did a little differently than the instructions.  The reasons for those changes are specified in added Markdown cells with header **Change**, located in the cell before the answer.  Additional notes, that I feel are worth emphasizing are placed in cells with header **Added Note**.  Actual code comments are still located in Pythyon cells

<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 15px; height: 80px">

# Project 1

### Building "Pokemon Stay"

---
You are an analyst at a "scrappy" online gaming company that specializes in remakes of last year's fads.

Your boss, who runs the product development team, is convinced that Pokemon Go's fatal flaw was that you had to actually move around outside. She has design mock-ups for a new game called Pokemon Stay: in this version players still need to move, but just from website to website. Pokemon gyms are now popular online destinations, and catching Pokemon in the "wild" simply requires browsing the internet for hours in the comfort of your home.

She wants you to program a prototype version of the game, and analyze the planned content to help the team calibrate the design.

#### Package imports

The pprint package below is the only package imported here, and it's not even strictly required to do any of the project. Printing python variables and objects with pprint can help to format them in a "prettier" way.

### Added Note : 
Because I use it later, I am also importing *numpy* in this cell.  I like to keep all my imports together.

In [1]:
from pprint import pprint
import numpy as np

<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 1. Defining a player

---

The player variables are:

    player_id : id code unique to each player (integer)
    player_name : entered name of the player (string)
    time_played : number of time played the game in minutes (float)
    player_pokemon: the player's captured pokemon (dictionary)
    gyms_visited: ids of the gyms that a player has visited (list)
    
Create the components for a player object by defining each of these variables. The dictionary and list variables should just be defined as empty; you can use any (correctly typed) values for the others.

In [2]:
# Initialize variable components of player class
# Set lists, dictionaries as empty values.  
# Give any value for other attributes  
 
player_id = 1
player_name = 'Ash Ketchum'
time_played = 45.5 #hours
player_pokemon = {}
gyms_visited = []

<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 2. Defining "gym" locations

---

As the sole programmer, Pokemon Stay will have to start small. To begin, there will be 10 different gym location websites on the internet. The gym locations are:

    1. 'reddit.com'
    2. 'amazon.com'
    3. 'twitter.com'
    4. 'linkedin.com'
    5. 'ebay.com'
    6. 'netflix.com'
    7. ~~'amazon.com'~~ -> 'gamefaqs.com
    8. 'stackoverflow.com'
    9. 'github.com'
    10. 'quora.com'

1. Set up a list of all the gym locations. This will be a list of strings.
2. Append two of these locations to your player's list of visited gyms.
3. Print the list.

### Change : 
*amazon.com* was listed twice in the list above.  Assuming this was a mistake, I changed the second one to a website of my choosing : *gamefaqs.com*.

In [3]:
# Delcare gym locations (websites), as listed above.
gym_locations = ['reddit.com', 'amazon.com', 'twitter.com', 'linkedin.com', 'ebay.com', 'netflix.com',
                'gamefaqs.com', 'stackoverflow.com', 'github.com', 'quora.com']


# Using numpy.random package.Create a list of two gyms, selected at random.
# List values do not repeat.
gyms_visited = list(np.random.choice(gym_locations, size=2, replace=False))

# Print results
pprint(gyms_visited)

['reddit.com', 'netflix.com']


<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 3. Create a pokedex

---

We also need to create some pokemon to catch. Each pokemon will be defined by these variables:

    pokemon_id : unique identifier for each pokemon (integer)
    name : the name of the pokemon (string)
    type : the category of pokemon (string)
    hp : base hitpoints (integer)
    attack : base attack (integer)
    defense : base defense (integer)
    special_attack : base special attack (integer)
    special_defense : base sepecial defense (integer)
    speed : base speed (integer)

We are only going to create 3 different pokemon with these `pokemon_id` and `pokemon_name` values:

    1 : 'charmander'
    2 : 'squirtle'
    3 : 'bulbasaur'

Create a dictionary that will contain the pokemon. The keys of the dictionary will be the `pokemon_id` and the values will themselves dictionaries that contain the other pokemon variables. The structure of the pokedex dictionary will start like so:
     
     {
         1: {
                 'name':'charmander',
                 'type':'fire',
                 ...
                 
The `type` of charmander, squirtle, and bulbasaur should be `'fire'`, `'water'`, and `'poison'` respectively. The other values are up to you, make them anything you like!

Print (or pretty print) the pokedex dictionary with the 3 pokemon.

### Change : 
I am using different values for the pokemon_id key/values.  This is so that I stay true to the Pokedex.
Corrected numbers are : 
- Bulbasaur : 1
- Charmander : 4
- Squirtle   : 7

### Added Note : 
Reading the *csv* file is for later, so here I am just inputing values manually.  For this we have to do it 1 by 1 since there's no way Python can guess what a pokemon's name is without explicitly saying it.

In [4]:
# Initialize the Pokedex as an empty dictionary
pokedex = {}

# Assign sub-dictionaries with pokemon attributes to pokemon_id keys in larger dictionary.

# Bulbasaur #1!!!
pokedex['1'] = {'name' : 'bulbasaur', 'type' : 'poison', 'hp' : 45, 'attack' : 49, 
                'defense' : 49, 'special_attack' : 65, 'special_defense' : 65, 'speed' : 45}
# Charmander
pokedex['4'] = {'name' : 'charmander', 'type' : 'fire', 'hp' : 39, 'attack' : 52, 
                'defense' : 43, 'special_attack' : 60, 'special_defense' : 50, 'speed' : 65}
# Squirtle 
pokedex['7'] = {'name' : 'squirtle', 'type' : 'water', 'hp' : 44, 'attack' : 48, 
                'defense' : 65, 'special_attack' : 50, 'special_defense' : 64, 'speed' : 43}

# Print results
pprint(pokedex)

{'1': {'attack': 49,
       'defense': 49,
       'hp': 45,
       'name': 'bulbasaur',
       'special_attack': 65,
       'special_defense': 65,
       'speed': 45,
       'type': 'poison'},
 '4': {'attack': 52,
       'defense': 43,
       'hp': 39,
       'name': 'charmander',
       'special_attack': 60,
       'special_defense': 50,
       'speed': 65,
       'type': 'fire'},
 '7': {'attack': 48,
       'defense': 65,
       'hp': 44,
       'name': 'squirtle',
       'special_attack': 50,
       'special_defense': 64,
       'speed': 43,
       'type': 'water'}}


<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 4. Create a data structure for players

---

### 4.1 

In order to maintain a database of multiple players, create a dictionary that keeps track of players indexed by `player_id`. 

The keys of the dictionary will be `player_id` and values will be dictionaries containing each player's variables (from question 1). 

Construct the `players` dictionary and insert the player that you defined in question 1, then print `players`.

### Added Note :
Although it does not explicitly state to do so, I am creating a function that will do this part for me.  This way I can simply use it again in the next section.

In [5]:
def init_player(player_id, player_name, time_played, player_pokemon, gyms_visited):
    ''' Construct a player dictionary.
    
    Arguments : player_id (int), player_name (string), time_played (float, in hours), player_pokemon (dict), 
    gyms visited (list).
    
    Dictionary keys are player_id's, for each value a sub-dictionary is constructed with key-value pairs of 
    all the player's attributes.
    '''
    
    # Construct dictionary, manually input all values.
    player_dict = {str(player_id) : 
                   {'player_name' : player_name,
                    'time_played' : time_played,
                    'player_pokemon' : player_pokemon,
                    'gyms_visited' : gyms_visited
                   }
                  }
    
    return player_dict

In [6]:
# Initialize player dictionary, and update it with values defined in Question 1
# Written in several steps for readability

# Create empty dictionary
players = {}

# Initialize player 1 using init_player function, defined above, and assign it to dummy variable.
_ = init_player(player_id, player_name, time_played, player_pokemon, gyms_visited)

# Update players dictionary with dummy dictionary
players.update(_)

# Print results
pprint(players)

{'1': {'gyms_visited': ['reddit.com', 'netflix.com'],
       'player_name': 'Ash Ketchum',
       'player_pokemon': {},
       'time_played': 45.5}}


---

### 4.2

Create a new player with `player_id = 2` in the `players` dictionary. Leave the `'player_pokemon'` dictionary empty. Append `'alcatraz'` and `'pacific_beach'` to the `'gyms_visited'` list for player 2.

The `'player_name'` and `'time_played'` values are up to you, but must be a string and float, respectively.

Remember, the player_id is the key for the player in the players dictionary.

Print the `players` dictionary with the new player inserted.

In [7]:
# Add second player to players dictionary
# Same steps as previous section (except creating the empty dictionary)
_ = init_player(2, 'Gary Oak', 48.6, {}, [])
players.update(_)

# Append alcatraz and pacific_beach to player 2's gyms_visited attribute
players['2']['gyms_visited'].append('alcatraz')
players['2']['gyms_visited'].append('pacific_beach')

# Print the updated dictionary
pprint(players)

{'1': {'gyms_visited': ['reddit.com', 'netflix.com'],
       'player_name': 'Ash Ketchum',
       'player_pokemon': {},
       'time_played': 45.5},
 '2': {'gyms_visited': ['alcatraz', 'pacific_beach'],
       'player_name': 'Gary Oak',
       'player_pokemon': {},
       'time_played': 48.6}}


<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 5. Add captured pokemon for each player

---

The `'player_pokemon'` keyed dictionaries for each player keep track of which of the pokemon each player has.

The keys of the `'player_pokemon'` dictionaries are the pokemon ids that correspond to the ids in the `pokedex` dictionary you created earlier. The values are integers specifying the stats for the pokemon.

Give player 1 a squirtle. Give player 2 charmander and a bulbasaur.

Print the players dictionary after adding the pokemon for each player.


In [8]:
# Give player 1 a Squirtle, player 2 a Charmander and Bulbasaur.
# Defined earlier, so we just have to assign correspoding values in dictionary.

players['1']['player_pokemon']['7'] = pokedex['7'] # Add Squirtle to player 1's pokemon (pokemon_id = 7)
players['2']['player_pokemon']['4'] = pokedex['4'] # Add Charmander to player 2's pokemon (pokemon_id = 4)
players['2']['player_pokemon']['1'] = pokedex['1'] # Add Bulbasaur to player 2's pokemon (pokemon_id = 1)

# Print results
pprint(players)

{'1': {'gyms_visited': ['reddit.com', 'netflix.com'],
       'player_name': 'Ash Ketchum',
       'player_pokemon': {'7': {'attack': 48,
                                'defense': 65,
                                'hp': 44,
                                'name': 'squirtle',
                                'special_attack': 50,
                                'special_defense': 64,
                                'speed': 43,
                                'type': 'water'}},
       'time_played': 45.5},
 '2': {'gyms_visited': ['alcatraz', 'pacific_beach'],
       'player_name': 'Gary Oak',
       'player_pokemon': {'1': {'attack': 49,
                                'defense': 49,
                                'hp': 45,
                                'name': 'bulbasaur',
                                'special_attack': 65,
                                'special_defense': 65,
                                'speed': 45,
                                'type': 'poison'},
       



## 6. What gyms have players visited?

---
<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">
### 6.1

Write a for-loop that:

1. Iterates through the `pokemon_gyms` list of gym locations you defined before.
2. For each gym, iterate through each player in the `players` dictionary with a second, internal for-loop.
3. If the player has visited the gym, print out "[player] has visited [gym location].", filling in [player] and [gym location] with the current player's name and current gym location.

#### ADDED NOTE : 
The loop only shows two gyms (for Ash Ketchum / `player : 1`), because `alcatraz` and `pacific_beach` were never added to the original list of gyms.  I am assuming that this was intended, since the point of the game is not to **GO** but to **STAY**.  Actually adding them would have been as simple as using list method extend() though.  I included another cell, with a dummy gym list in case this was not the case.

In [9]:
# Loop through gym locations (pokemon_gyms)
for gym in gym_locations:
    
    # Iterate through players in players dictionary
    for p in players:
        
        # Check if gym (item of pokemon_gyms) is in each player's dictionary using an if statement
        if(gym in players[p]['gyms_visited']):
            
            # Create formated text, done in separate steps than print statements for readability
            text = '{} has visited {}'.format(players[p]['player_name'], gym)
            print(text)

Ash Ketchum has visited reddit.com
Ash Ketchum has visited netflix.com


In [10]:
# Do same as above, but with dummy list of gyms.  Make sure we are making a copy when declaring so
# that we don't change the original one.

# Create dummy list
dummy_list = list(gym_locations)
dummy_list.extend(['alcatraz', 'pacific_beach'])

# Run code from previous cell
# Loop through gym locations (pokemon_gyms)
for gym in dummy_list:
    
    # Iterate through players in players dictionary
    for p in players:
        
        # Check if gym (item of pokemon_gyms) is in each player's dictionary using an if statement
        if(gym in players[p]['gyms_visited']):
            
            # Create formated text, done in separate steps than print statements for readability
            text = '{} has visited {}'.format(players[p]['player_name'], gym)
            print(text)

Ash Ketchum has visited reddit.com
Ash Ketchum has visited netflix.com
Gary Oak has visited alcatraz
Gary Oak has visited pacific_beach


<img src="http://imgur.com/xDpSobf.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">
### 6.2

How many times did that loop run? If you have N gyms and also N players, how many times would it run as a function of N?

Can you think of a more efficient way to accomplish the same thing? 

(You can write your answer as Markdown text.)

#### Answer : 
Generally speaking, the total number of iterations, $N$ of a nested loop is :

\begin{equation} 
N = N_i \times N_j 
\end{equation}

where, $N_j$ is the number of iterations of the child loop, and $N_i$ is the number of iterations of the parent loop.  In the case of $N$ gyms and $N$ players, we have : 

\begin{equation}
N = N \times N = N^2
\end{equation}

In our case, we have a list of 10 gyms, and 2 players, so the answer would be : 

\begin{equation}
N_{runs} = 10 \times 2 = 20
\end{equation}

To check this, we can run the same loops ony replacing the content with a counter.

In [11]:
# Initialize counter 
counter = 0

# Repeat previous loop, but replacing the content with the counter
for gym in gym_locations:
    for p in players:
        # Increment counter
        counter += 1

# Print counter results
print(counter)

20


We indeed, have a total of 20 iterations.

<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 7. Calculate player "power".

---

Define a function that will calculate a player's "power". Player power is defined as the sum of the base statistics all of their pokemon.

Your function will:

1. Accept the `players` dictionary, `pokedex` dictionary, and a player_id as arguments.
2. For the specified player_id, look up that player's pokemon and their level(s).
3. Find and aggregate the attack and defense values for each of the player's pokemon from the `pokedex` dictionary.
4. Print "[player name]'s power is [player power].", where the player power is the sum of the base statistics for all of their pokemon.
5. Return the player's power value.

Print out the pokemon power for each of your players.

### Change :
It seemed odd to me to only use the Pokemon's Attack and Defense as a basis of player power.  So I am instead summing all stats.

In [12]:
# Function to calculate player Power
# Note : I am also adding the special attack/defense and speed to this calculations
def player_power(players, pokedex, player_id):
    ''' Return player power.
    
    Sums base stats of all pokemon in player's dictionary.
    Arguments : players (dict), pokedex (dict), player_id (int or str)
    '''
    
    # Convert player_id to string
    player_id = str(player_id)
    
    # Start by getting the player's pokemon dictionary (the ones he has)
    player_pk = players[player_id]['player_pokemon']
    
    # Initialize player power as zero
    power = 0
    
    # Loop through key, value pairs in player_pk dictionary
    # Each value is another dictionary with the pokemon's stats, so we need 
    # to embed another for-loop that iterates through those stats
    for pokemon_id, stat_dict in player_pk.items():
        for stat_name, stat_value in stat_dict.items():
            
            # Get all attributes except name and type, since those don't make sense in the context.
            if(stat_name != 'name' and stat_name != 'type'):
                # Increment power 
                power += stat_value
    
    # Print results, and return player power
    pprint("{}'s power is {}".format(players[player_id]['player_name'], power))
    return power
    
    
    

In [13]:
player_power(players, pokedex, 2)

"Gary Oak's power is 627"


627

<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 8. Load a pokedex file containing all the pokemon

---

### 8.1

While you were putting together the prototype code, your colleagues were preparing a dataset of Pokemon and their attributes. (This was a rush job, so they may have picked some crazy values for some...)

The code below loads information from a comma separated value (csv) file. You need to parse this string into a more useable format. The format of the string is:

- Rows are separated by newline characters: \n
- Columns are separated by commas: ,
- All cells in the csv are double quoted. Ex: "PokedexNumber" is the first cell of the first row.


Using for-loops, create a list of lists where each list within the overall list is a row of the csv/matrix, and each element in that list is a cell in that row. Additional criteria:

1. Quotes are removed from each cell item.
2. Numeric column values are converted to floats.
3. There are some cells that are empty and have no information. For these cells put a -1 value in place.

Your end result is effectively a matrix. Each list in the outer list is a row, and the *j*th elements of list together form the *j*th column, which represents a data attribute. The first three lists in your pokedex list should look like this:

    ['PokedexNumber', 'Name', 'Type', 'Total', 'HP', 'Attack', 'Defense', 'SpecialAttack', 'SpecialDefense', 'Speed']
    [1.0, 'Bulbasaur', 'GrassPoison', 318.0, 45.0, 49.0, 49.0, 65.0, 65.0, 45.0]
    [2.0, 'Ivysaur', 'GrassPoison', 405.0, 60.0, 62.0, 63.0, 80.0, 80.0, 60.0]

### Added Note : 
I wasn't sure if the directions mean't that we were *only* allowed to use for loops, meaning no python string methods.  So this next part is done entirely with for-loops and if-else statements.  I basically loop over each character in the raw_pd string, and have different behaviors depending on the type of character.  I also assumed that we were not allowed to change the part of the code that reads the file, meaning we can't just have it read as a .csv, and stored accordingly.  Definitely not the easiest way to do things, considering what's available to us, but it was good practice.

In the next cell I define a function which helps me convert characters/strings to appropriate values, as indicated above.

In [14]:
def check_string(string):
    ''' Check if string is a number or an empty value.
    
    if number : return value as float
    if empty  : return -1 as int
    otherwise, just return the original string
    '''
    if(string == ''):
        return -1
    
    elif(string.isdigit()): 
        return float(string)
    
    else:
        return string

In [17]:
# Code to read in pokedex info
raw_pd = ''
pokedex_file = 'pokedex_basic.csv'
with open(pokedex_file, 'r') as f:
    raw_pd = f.read()

# Splitting into lists of lists using a for-loop (which is not the easiest way to do this)


# Initialize empty nested list that will store the final product.
# Also, initialize row and column counters to structure the new list.
pkdx = [['']]
row, col  = 0, 0

# Loop over each letter in raw_pd
for i in range(len(raw_pd)):
    
    # Store letter in variable (for clarity)
    letter = raw_pd[i]
    
    if(letter == '"'):
        # If the letter is a quotation mark ignore it and go to next iteration.
        continue
        
    elif(letter == '\n'):
        # If it is a new line, check/convert the latest value in list using check_string
        pkdx[row][col] = check_string(pkdx[row][col])
        
        # Create new empty sub-list, increment row number,
        # reset column number (new row, start at 0),
        # and continue to next iteration.
        pkdx.append([''])
        row += 1
        col  = 0
        continue
    
    elif(letter == ','):
        # If letter is a comma, we are going to a new list element, we need to
        # check for digits, empty values, increment column number, and append
        # a new empty string for next item.
        pkdx[row][col] = check_string(pkdx[row][col])
        col += 1
        pkdx[row].append('')
        continue
        
    # If we've gotten passed the if-else statements, that means we are actually looking 
    # at a letter and we need to add it to the string in the current row-col
    pkdx[row][col] += letter

    # We need the loop to behave differently if we are on the last letter.
    # We need to explicitly check the last list element.
    if i == len(raw_pd)-2:
        # Bad coding : I explicitly specified - 2 because I knew that the last character is a quotation mark.
        # Not sure how to go around this issue.
        pkdx[row][col] = check_string(pkdx[row][col])
    
# Print 1st 5 Results
for entry in pkdx[:5]:
    pprint(entry)

['PokedexNumber',
 'Name',
 'Type',
 'Total',
 'HP',
 'Attack',
 'Defense',
 'SpecialAttack',
 'SpecialDefense',
 'Speed']
[1.0, 'Bulbasaur', 'GrassPoison', 318.0, 45.0, 49.0, 49.0, 65.0, 65.0, 45.0]
[2.0, 'Ivysaur', 'GrassPoison', 405.0, 60.0, 62.0, 63.0, 80.0, 80.0, 60.0]
[3.0, 'Venusaur', 'GrassPoison', 525.0, 80.0, 82.0, 83.0, 100.0, 100.0, 80.0]
[3.0,
 'VenusaurMega Venusaur',
 'GrassPoison',
 625.0,
 80.0,
 100.0,
 123.0,
 122.0,
 120.0,
 80.0]


#### ADDED NOTE : 
<br>
Using a for loop in the way we just did is definitely not the easiest way to do this... As a demonstration, I am going to do the same thing below, but this time using string methods.

In [19]:
# Assign raw_pd to new variable and remove quotation marks 
pkdx_easy = raw_pd.replace('\"', '')

# Split list on end line characters to separate each row
pkdx_easy = pkdx_easy.split('\n')

# Split each row in the new list into it's columns, separated by commas
pkdx_easy = [r.split(",") for r in pkdx_easy]

# Now all we have to do is convert digit strings into floats, and replace empty values with 
# -1.  I guess I have to use a loop in the end.
pkdx_easy = [[float(c) if c.isdigit() else -1 if c == '' else c for c in r] for r in pkdx_easy]

# Print 1st 5 results
for entry in pkdx_easy[:5]:
    pprint(entry)

['PokedexNumber',
 'Name',
 'Type',
 'Total',
 'HP',
 'Attack',
 'Defense',
 'SpecialAttack',
 'SpecialDefense',
 'Speed']
[1.0, 'Bulbasaur', 'GrassPoison', 318.0, 45.0, 49.0, 49.0, 65.0, 65.0, 45.0]
[2.0, 'Ivysaur', 'GrassPoison', 405.0, 60.0, 62.0, 63.0, 80.0, 80.0, 60.0]
[3.0, 'Venusaur', 'GrassPoison', 525.0, 80.0, 82.0, 83.0, 100.0, 100.0, 80.0]
[3.0,
 'VenusaurMega Venusaur',
 'GrassPoison',
 625.0,
 80.0,
 100.0,
 123.0,
 122.0,
 120.0,
 80.0]


Same results, much clearer and concise.

<img src="http://imgur.com/xDpSobf.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

### 8.2 Parse the raw pokedex with list comprehensions

---

Perform the same parsing as above, but **using only a single list comprehension** instead of for loops. You may have nested list comprehensions within the main list comprehension! The output should be exactly the same.

#### Added Note 
This is essentially the same thing as in my demonstration in the previous section, but done in a single statement.  Spatially efficient, but not that readable.  I am also adding a bit of code because I noticed that the Mega evolutions of the pokemon share the same ID numbers as their counterparts. Because of this, I will add 0.1 to the Mega Evolution Y ID number, and 0.2 for the Mega Evolution X (3 identical IDs) before moving on to the next section.

In [21]:
pokedex = [[float(c) if c.isdigit() else -1 if c == '' else c for c in _.split(',')]
             for _ in raw_pd.replace('\"', '').split("\n")]

In [22]:
# Correction for repeated pokemon id's
# Add 0.1 to pokedex number of mega evolutions X, 0.2 to mega evolutions Y.  
# I'm doing this one quick and dirty.
for i,x in enumerate(pokedex[1:], 1):
    if(i == 1):
        continue
    
    if(x[0] == pokedex[i-1][0]):
        pokedex[i][0] += 0.1
    
    elif (x[0] == pokedex[i-1][0] - 0.1):
        pokedex[i][0] += 0.2

# Print 1st 5
for entry in pokedex[:5]:
    pprint(entry)

['PokedexNumber',
 'Name',
 'Type',
 'Total',
 'HP',
 'Attack',
 'Defense',
 'SpecialAttack',
 'SpecialDefense',
 'Speed']
[1.0, 'Bulbasaur', 'GrassPoison', 318.0, 45.0, 49.0, 49.0, 65.0, 65.0, 45.0]
[2.0, 'Ivysaur', 'GrassPoison', 405.0, 60.0, 62.0, 63.0, 80.0, 80.0, 60.0]
[3.0, 'Venusaur', 'GrassPoison', 525.0, 80.0, 82.0, 83.0, 100.0, 100.0, 80.0]
[3.1,
 'VenusaurMega Venusaur',
 'GrassPoison',
 625.0,
 80.0,
 100.0,
 123.0,
 122.0,
 120.0,
 80.0]


<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 9. Write a function to generate the full pokedex

---

Write a function that recreates the pokedex you made before, but with the data read in from the full pokemon file. The `PokedexNumber` should be used as the `pokemon_id` key values for the dictionary of pokemon.

Your function should:

1. Take the parsed pokedex information you created above as an argument.
2. Return a dictionary in the same format as your original pokedex you created before containing the information from the parsed full pokedex file.

To test the function, print out the pokemon with id = 100.

### Change :
I've kept all the pokemon key's (ID numbers as floats).  This is because of my correction of repeated ID numbers, since I'm using decimals.  So testing the function will have id = 100.0 instead of id = 100.

In [20]:
# Define function
def make_pokedex_dict(pokemon_info):
    '''
        Convert pokemon stats list of lists into a dictionary with pokemon_ids as keys.
    '''
    
    # We have to make sure that we are using a copy of the list for manipulation.
    # This way we don't change the original one.
    pokedex = list(pokemon_info)

    # We know that the first row in the pokedex is a list of names for the values.
    # Extract it and delete it from list. Also making sure they are all strings since they will
    # used as dictionary keys
    keys = [str(p) for p in pokedex[0]]
    del(pokedex[0])
 
    # Now the fun part! Creating the dictionary
    
    # Start by initializing an empty dictionary, this will be the one returned. 
    # We will update it as we create the sub-dictionaries.
    pokedex_dict = {}

    # Loop throuh sub-lists 
    for pokemon in pokedex :
        
        # Create subdictionary, and update the big one
        pokemon_dict = {str(keys[k]) : v for k,v in enumerate(pokemon)}
        pokemon_dict = {str(pokemon_dict['PokedexNumber']) : pokemon_dict}
        pokedex_dict.update(pokemon_dict)

    return pokedex_dict
    
# make_pokemon_dict(pokedex)
pokedex_dict = make_pokedex_dict(pokedex)

# Print test pokemon
pprint(pokedex_dict['100.0'])

{'Attack': 30.0,
 'Defense': 50.0,
 'HP': 40.0,
 'Name': 'Voltorb',
 'PokedexNumber': 100.0,
 'SpecialAttack': 55.0,
 'SpecialDefense': 55.0,
 'Speed': 100.0,
 'Total': 330.0,
 'Type': 'Electric'}


<img src="http://i.imgur.com/GCAf1UX.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 10. Write a function to generate a "filtered" pokedex
---
Your function should:
1. Take the parsed pokedex information you created above as an argument.
1. Take a dictionary as a parameter with keys matching the features of the Pokedex, filtering by exact match for string type values, and/or filter continuous variables specified value that is greater than or equal to the dictionary key parameter.
1. Return multiple elements from the Pokedex

Example:

```python

# Only filter based on parameters passed
filter_options = {
    'Attack':   25,
    'Defense':  30,
    'Type':     'Electric'
}

# Return records with attack >= 24, defense >= 30, and type == "Electric"
# Also anticipate that other parameters can also be passed such as "SpecialAttack", "Speed", etc.
filtered_pokedex(pokedex_data, filter=filter_options)

# Example output:
# [{'Attack': 30.0,
#  'Defense': 50.0,
#  'HP': 40.0,
#  'Name': 'Voltorb',
#  'SpecialAttack': 55.0,
#  'SpecialDefense': 55.0,
#  'Speed': 100.0,
#  'Total': 330.0,
#  'Type': 'Electric'},
#  {'Attack': 30.0,
#  'Defense': 33.0,
#  'HP': 32.0,
#  'Name': 'Pikachu',
#  'SpecialAttack': 55.0,
#  'SpecialDefense': 55.0,
#  'Speed': 100.0,
#  'Total': 330.0,
#  'Type': 'Electric'},
#  ... etc
#  ]

```



In [21]:
def filtered_pokedex(pokedex, filter_options={}):
    ''' Return filtered pokedex.
    
    Construct a dictionary of pokemon whose stats are greater or equal than the specified values.
    
    Argument   : pokedex (dict)
    Parameters : filter_options, passed as dictionary of stat-key : stat-value pairs.
    
    Types, Names, and PokedexNumber can also be specified in filter_options.
    If filter_options is not specified, function returns the entire pokedex.
    
    Basically, we are flagging any key in the dictionary that does not meet the 
    criteria.  Flagged key-value pairs are subsequentially deleted from a copy of 
    the pokedex.  The reduced dictionary is then returned.
    '''
    
    # Make sure only dictionaries are being entered, other wise return TypeError
    if (isinstance(pokedex, dict) == False or isinstance(filter_options, dict) == False) :
        return TypeError
    
    # Check if filter options are specified, if not return original dictionary
    if (filter_options == {}):
        return pokedex    
    
    # Make copy of pokedex for manipulation, this is the one that will be returned at the end.
    filtered_pokedex = pokedex.copy()
    
    # For this I will loop over all elements of dictionary, and flag keys that need to be
    # deleted by adding them to a list
    flagged = []
    
    # Loop over all key-value pairs in dictionary and flag them if they 
    # do not meet the filter requirements
    for pokedex_number, pokemon in filtered_pokedex.items(): 
        flag = False
        for filter_key, filter_value in filter_options.items():
            
            # Needs different behaviours depending on type of values (strings vs floats)
            if isinstance(pokemon[filter_key], str):
                if pokemon[filter_key] != filter_value:
                    flag = True
                    break
            elif(isinstance(pokemon[filter_key], float)):
                if(pokemon[filter_key] < filter_value) :
                    flag = True
                    break

        if (flag == True):
            flagged.append(pokedex_number)
            continue
    
    # Delete elements in flagged list
    for f in flagged : 
        del(filtered_pokedex[f])
    
    return filtered_pokedex

In [25]:
# Test function
pprint(filtered_pokedex(pokedex_dict, filter_options={'Attack' : 100, 'Speed' : 100,
                                                      'SpecialAttack' : 100, 'Type' : 'Psychic'}))

{'150.0': {'Attack': 110.0,
           'Defense': 90.0,
           'HP': 106.0,
           'Name': 'Mewtwo',
           'PokedexNumber': 150.0,
           'SpecialAttack': 154.0,
           'SpecialDefense': 90.0,
           'Speed': 130.0,
           'Total': 680.0,
           'Type': 'Psychic'},
 '150.2': {'Attack': 150.0,
           'Defense': 70.0,
           'HP': 106.0,
           'Name': 'MewtwoMega Mewtwo Y',
           'PokedexNumber': 150.2,
           'SpecialAttack': 194.0,
           'SpecialDefense': 120.0,
           'Speed': 140.0,
           'Total': 780.0,
           'Type': 'Psychic'},
 '151.0': {'Attack': 100.0,
           'Defense': 100.0,
           'HP': 100.0,
           'Name': 'Mew',
           'PokedexNumber': 151.0,
           'SpecialAttack': 100.0,
           'SpecialDefense': 100.0,
           'Speed': 100.0,
           'Total': 600.0,
           'Type': 'Psychic'},
 '386.1': {'Attack': 180.0,
           'Defense': 20.0,
           'HP': 50.0,
           


## 9. Descriptive statistics on the prototype pokedex

<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">
### 9.1

What is the population mean and standard deviation of the "Total" attribute for all characters in the Pokedex?



### Added Note : 
For the sake of simplicity I will be using numpy functions for this one.

In [86]:
# Perhaps there's a more direct way to do this with dictionaries, but this is what I've got :
# Put all 'Total' values into a list, and use numpy to get mean and standard deviation.
total_mean = np.mean([v['Total'] for d,v in pokedex_dict.items()])
total_std  = np.std([v['Total']  for d,v in pokedex_dict.items()])
print('Total, mean : {}\nTotal, standard deviation : {}'.format(total_mean, total_std))

Total, mean : 434.75188916876573
Total, standard deviation : 120.12531304418118


The mean of the Total attribute for all pokemon is : 434.75 <br>The standard deviation of the Total attribute for all pokemon is : 120.13

<img src="http://imgur.com/l5NasQj.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">
### 9.2

The game is no fun if the characters are wildly unbalanced! Are any characters "overpowered", which we'll define as having a "Total" more than three standard deviations from the population mean?

In [88]:
# Get the limit above which characters are considered overpowered : 
over9000 = total_mean + 3 * total_std
print('Pokemon with Total attribute over {}, are overpowered'.format(over9000))

Pokemon with Total attribute over 795.1278283013093, are overpowered


In [90]:
# Now I'm just going to do a simple for loop, and print out the name and key of any overpowered pokemon.
print('Overpowered pokemon are :')
for pokedexNumber, pokemon in pokedex_dict.items():
    if(pokemon['Total'] > over9000):
        print('{} : {}, Total = {}'.format(pokemon['PokedexNumber'], pokemon['Name'], pokemon['Total']))

Overpowered pokemon are :
150.1 : MewtwoMega Mewtwo X, Total = 800.0


I guess Mega Mewtwo X is the only over powered pokemon.
<img src="https://cdn.bulbagarden.net/upload/7/7f/150Mewtwo-Mega_X.png" style="float: center; margin: 25px 15px 50px 0px; height: 327px"> 

<img src="http://imgur.com/xDpSobf.png" style="float: left; margin: 25px 15px 0px 0px; height: 25px">

## 10. Calibrate the frequency of Pokemon

The design team wants you to make the powerful Pokemon rare, and the weaklings more common. How would you set the probability $p_i$ of finding Pokemon *i* each time a player visits a gym?

Write a function that takes in a Pokedex number and returns a value $p_i$ for that character.

Hint: there are many ways you could do this. What do _you_ think makes sense? Start with simplifying assumptions: for example, you could assume that the probabilities of encountering any two Pokemon on one visit to a gym are independent of each other.

#### Answer :

I suppose I could makeup a function, $p_i(T)$, which gives the probability as a function of the Total attribute.  Keeping in mind that the probablity has to be between 0 and 1, here is my completely made up function : 

\begin{equation}
p_i(T) = 1 - \frac{T}{T_{max} + 0.001}
\end{equation}

where $T_{max}$ is the highest value of $T$ in the pokedex (a.k.a Mega Mewtwo X).  The $+0.001$ is so that Mega Mewtwo still has a chance to appear.

### Change : 
I'm not sure how it's supposed to only take a pokemon's number, since the dictionary shouldn't be part of the function's scope.  Same thing for what is the maximum Total value.  So instead it's taking a pokemon's number and and the entire pokedex.  This is actually good because, then we can even have different distributions for filtered pokedexes.  While I'm at it, I'm adding the option to search by ID or Name.  This means that I'm passing the ID as
a parameter.

In [92]:
# Here is the function in practice
def get_pokemon_probablity(pokedex_dict, pokemon_id='1.0', pokemon_name = ''):
    ''' Return probablity of pokemon appearing at a gym.
    '''
    
    # If pokemon_name is passed, match it to it's ID
    if (pokemon_name != ''):
        for k,v in pokedex_dict.items():
            if(v['Name'] == pokemon_name):
                pokemon_id = k

                
    # Make sure ID number is a string :
    pokemon_id = str(pokemon_id)
    
    # Get maximum Total value
    Tmax = max([v['Total'] for d,v in pokedex_dict.items()])
    
    # Extract pokemon's Total value
    T = pokedex_dict[pokemon_id]['Total']
    
    # Get probability
    p_i = 1 - T / (Tmax + 0.001)
    
    print('Probability of finding {} (ID = {}) is {}'.format(pokedex_dict[pokemon_id]['Name'], pokemon_id, p_i))
    
    return p_i

In [94]:
# Test function
get_pokemon_probablity(pokedex_dict, 150.1)
get_pokemon_probablity(pokedex_dict, pokemon_name='Caterpie')

# Check to see if max probability is still <= 1
tmin = min([v['Total'] for d,v in pokedex_dict.items()])
for k,v in pokedex_dict.items():
    if (v['Total'] == tmin):
        print(k)
get_pokemon_probablity(pokedex_dict, 191.0)


Probability of finding MewtwoMega Mewtwo X (ID = 150.1) is 1.249998437424793e-06
Probability of finding Caterpie (ID = 10.0) is 0.7562503046871192
191.0
Probability of finding Sunkern (ID = 191.0) is 0.7750002812496484


0.7750002812496484